pkg/api/vm/handler.go

```diff
@@ -633,15 +685,54 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
 		return fmt.Errorf("failed to serialize payload,: %v", err)
 	}
 	// Ref: https://kubevirt.io/api-reference/v0.44.0/operations.html#_v1vm-removevolume
-	return h.virtSubresourceRestClient.
+	if err = h.virtSubresourceRestClient.
 		Put().
 		Namespace(namespace).
 		Resource(vmResource).
 		Name(name).
 		SubResource(strings.ToLower(removeVolume)).
 		Body(body).
 		Do(ctx).
-		Error()
+		Error(); err != nil {
+		return err
+	}
+
+	// Get VM object again, because we've updated it.
+	vm, err = h.vmCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	if pvcName != "" {
+		var volumeClaimTemplates []corev1.PersistentVolumeClaim
+		vmCopy := vm.DeepCopy()
+		anno := vmCopy.GetAnnotations()
+		if volumeClaimTemplatesJSON, ok := anno[util.AnnotationVolumeClaimTemplates]; ok {
+			if err = json.Unmarshal([]byte(volumeClaimTemplatesJSON), &volumeClaimTemplates); err != nil {
+				return fmt.Errorf("failed to unserialize %s, error: %v", util.AnnotationVolumeClaimTemplates, err)
+			}
+		}
+		for i, volumeClaimTemplate := range volumeClaimTemplates {
+			if volumeClaimTemplate.Name == pvcName {
+				volumeClaimTemplates[i], volumeClaimTemplates[len(volumeClaimTemplates)-1] = volumeClaimTemplates[len(volumeClaimTemplates)-1], volumeClaimTemplates[i]
+				volumeClaimTemplates = volumeClaimTemplates[:len(volumeClaimTemplates)-1]
+				break
+			}
+		}
+		volumeClaimTemplatesJSON, err := json.Marshal(volumeClaimTemplates)
+		if err != nil {
+			return fmt.Errorf("failed to serialize payload %v, error: %v", volumeClaimTemplates, err)
+		}
+		anno[util.AnnotationVolumeClaimTemplates] = string(volumeClaimTemplatesJSON)
+		vmCopy.SetAnnotations(anno)
+		if !reflect.DeepEqual(vm, vmCopy) {
+			if _, err = h.vms.Update(vmCopy); err != nil {
+				return fmt.Errorf("failed to update vm %s/%s, error: %v", namespace, name, err)
+			}
+		}
+	}
+
+	return nil
 }
 
```

Comment:
we can add a common function for addVolume and removeVolume, e.g:

```go
func (h *vmActionHandler) UpdateVMVolumeClaimTemplate(vm *kubevirtv1.VirtualMachine, updateVolumeClaimTemplate func([]corev1.PersistentVolumeClaim) ([]corev1.PersistentVolumeClaim, bool)) error {
	var volumeClaimTemplates []corev1.PersistentVolumeClaim
	anno := vm.GetAnnotations()
	if volumeClaimTemplatesJSON, ok := anno[util.AnnotationVolumeClaimTemplates]; ok {
		if err := json.Unmarshal([]byte(volumeClaimTemplatesJSON), &volumeClaimTemplates); err != nil {
			return fmt.Errorf("failed to unserialize %s, error: %v", util.AnnotationVolumeClaimTemplates, err)
		}
	}

	var changed bool
	volumeClaimTemplates, changed = updateVolumeClaimTemplate(volumeClaimTemplates)
	if !changed {
		return nil
	}

	volumeClaimTemplatesJSON, err := json.Marshal(volumeClaimTemplates)
	if err != nil {
		return fmt.Errorf("failed to serialize payload %v, error: %v", volumeClaimTemplates, err)
	}
	anno[util.AnnotationVolumeClaimTemplates] = string(volumeClaimTemplatesJSON)
	vmCopy := vm.DeepCopy()
	vmCopy.SetAnnotations(anno)
	if !reflect.DeepEqual(vm, vmCopy) {
		if _, err = h.vms.Update(vmCopy); err != nil {
			return fmt.Errorf("failed to update vm %s/%s, error: %v", vm.Namespace, vm.Name, err)
		}
	}
	return nil
}
```

after that, we can simplify the code:
```go
addVolumeClaimTemplate := func(volumeClaimTemplates []corev1.PersistentVolumeClaim) ([]corev1.PersistentVolumeClaim, bool) {
		for _, volumeClaimTemplate := range volumeClaimTemplates {
			if volumeClaimTemplate.Name == input.VolumeSourceName {
				return volumeClaimTemplates, false
			}
		}
		volumeClaimTemplates = append(volumeClaimTemplates, corev1.PersistentVolumeClaim{
			ObjectMeta: metav1.ObjectMeta{
				Name: pvc.Name,
			},
			Spec: corev1.PersistentVolumeClaimSpec{
				AccessModes:      pvc.Spec.AccessModes,
				Resources:        pvc.Spec.Resources,
				VolumeMode:       pvc.Spec.VolumeMode,
				StorageClassName: &pvc.Spec.VolumeName,
			},
		})
		return volumeClaimTemplates, true
	}

	return h.UpdateVMVolumeClaimTemplate(vm, addVolumeClaimTemplate)
```
and
```go
if pvcName != "" {
		removeVolumeClaimTemplate := func(volumeClaimTemplates []corev1.PersistentVolumeClaim) ([]corev1.PersistentVolumeClaim, bool) {
			for i, volumeClaimTemplate := range volumeClaimTemplates {
				if volumeClaimTemplate.Name == pvcName {
					volumeClaimTemplates[i], volumeClaimTemplates[len(volumeClaimTemplates)-1] = volumeClaimTemplates[len(volumeClaimTemplates)-1], volumeClaimTemplates[i]
					volumeClaimTemplates = volumeClaimTemplates[:len(volumeClaimTemplates)-1]
					return volumeClaimTemplates, true
				}
			}
			return volumeClaimTemplates, false
		}
		return h.UpdateVMVolumeClaimTemplate(vm, removeVolumeClaimTemplate)
	}
```